design/orchestrators.md

```diff
@@ -0,0 +1,231 @@
+passing tasks to the update supervisor by slot, it groups them by node. This
+means rolling updates will go node-by-node instead of slot-by-slot.
+
+## Restart supervisor
+
+The restart supervisor manages the process of shutting down a task, and possibly
+starting a replacement task. Its entry point is a `Restart` method which is
+called inside a store write transaction. It atomically changes the state of the
+old task to `Shutdown`, and, if it's appropriate to start a replacement task,
+creates a new task in the same slot (replicated service) or on the same node
+(global service).
+
+If the service is set up with a restart delay, the restart supervisor handles
+this delay too. It initially creates the new task with the desired state
+`Ready`, and only changes the desired state to `Running` after the delay has
+elapsed. One of the things the orchestrators do when they start up is check for
+tasks that were in this delay phase of being restarted, and make sure they get
+advanced to `Running`.
+
+In some cases, a task can fail or be rejected before its desired state reaches
+`Running`. One example is a failure to pull an image from a registry. The
+restart supervisor tries to make sure this doesn't result in fast restart loops
+that effectively ignore the restart delay. If `Restart` is called on a task that
+the restart supervisor is still in the process of starting up - i.e. it hasn't
+moved the task to `Running` yet - it will wait for the restart delay to elapse
+before triggering this second restart.
+
+The restart supervisor implements the logic to decide whether a task should be
+restarted, and since this can be dependent on restart history (when
+`MaxAttempts`) is set, the restart supervisor keeps track of this history. The
+history isn't persisted, so some restart behavior may be slightly off after a
+restart or leader election.
+
+Note that a call to `Restart` doesn't always end up with the task being
+restarted - this depends on the service's configuration. `Restart` can be
+understood as "make sure this task gets shut down, and maybe start a replacement
+if the service configuration says to".
+
+## Update supervisor
+
+The update supervisor is the component that updates existing tasks to match the
+latest version of the service. This means shutting down the old task and
+starting a new one to replace it. The update supervisor implements rolling
+updates and automatic rollback.
+
+The update supervisor operates on an abstract notion of slots, which are either
+slot numbers for replicated services, or node IDs for global services. You can
+think of it as reconciling the contents of each slot with the service. If a slot
+has more than one task or fewer than one task, it corrects that. If the task (or
+tasks) in a slot are out of date, they are replaced with a single task that's up
+to date.
+
+Every time the update supervisor is called to start an update, it spawns an
```

Comment:
Should we note that updates are called on a particular service, and there is only 1 updater per service at any given time?